feat: breakage report images

Ben754444 opened the pull request:
### Feature summary

Support uploading images with breakage reports

### Work included

- Added file upload to report form
- Images displayed on the 'view breakage' page
- New table to store references to images 
- New route for streaming the images (`/equipment/repairs/:reportId/images/:imageId`)

### Points for discussion 

- Should an image be required for a report to be submitted?
- What can a user do if they upload an image and wish for it to be removed? (should a delete button be added?)
- How should the images be included in the email to E&S?
  - (probably best not to)
- Should we support short videos?

bnjns commented:
My word's not gospel, but my thoughts:

> Should an image be required for a report to be submitted?

Personally I'd say keep them optional, but this is probably up to the E&Ses to decide - keep in mind that you don't want to make it too difficult to report breakages, or people won't use it :wink: 

> What can a user do if they upload an image and wish for it to be removed? (should a delete button be added?)

Good question :thinking: I'd err on the side of "wait until this becomes a problem" so that the PR doesn't become too large. Worst case, we can do this manually.

> How should the images be included in the email to E&S?

I can't remember what the email includes - if it's just a link to the report I'd leave as is, otherwise maybe keep it simple and just have a count of the number of images included (or a simple yes/no)? You could attach them if you wanted, but I've not done that with Laravel before so not entirely sure how that works (and you may run into attachment size restrictions).

> Should we support short videos?

I vote no :stuck_out_tongue:  and definitely not in the same PR!

Ben754444 commented:
Those were pretty much my thoughts, we can check with E&S about optionality once they see it on staging. I'm going to do a quick clean up then should be good for a review

bnjns commented on app/Http/Controllers/Equipment/RepairsController.php:
```diff
@@ -63,14 +64,28 @@ public function store(RepairRequest $request)
+        if ($request->hasFile('images')) {
+            $images = $request->file('images');
+            $i = 0;
+            foreach ($images as $image) {
```
You could probably use an indexed foreach here, to save having to explicitly declare `$i`:

```php
foreach ($images as $i => $image) {
```

Ben754444 replied:
Dropped the `position_id` so it won't be necessary to keep track of the index

bnjns commented on app/Models/Equipment/Breakage.php:
```diff
@@ -57,4 +57,22 @@ public function user()
+    public function saveImage($image)
+    {
+        // save image to /resources/breakages
+        $filename = uniqid() . '.' . $image->extension();
```
I would recommend using the image's ID here - it will make it much easier to trace in the filesystem, compared to a random name. Plus, you then don't need to explicitly save this in the DB.

bnjns commented on app/Models/Equipment/BreakageImage.php:
```diff
@@ -0,0 +1,60 @@
+    protected $fillable = [
+        'filename',
+        'mime',
+        'report_id',
```
nit: it's generally recommended that FK names use the table name, eg `equipment_breakage_id`

bnjns commented on app/Models/Equipment/Breakage.php:
```diff
@@ -57,4 +57,22 @@ public function user()
+     */
+    public function images()
+    {
+        return $this->hasMany('App\Models\Equipment\BreakageImage', 'report_id');
```
Do we want to sort this by the position? Otherwise, we probably don't need to store the position at all.

bnjns commented on chart/templates/deployment.yaml:
```diff
@@ -30,6 +30,8 @@ spec:
               mountPath: /var/www/resources/elections
             - name: resources
               mountPath: /var/www/resources/resources
+            - name: breakages
+              mountPath: /var/www/resources/breakages
```
I'm not sure why this isn't being flagged by the validation check, but you'll need to add a volume to the `volumes` section below.

bnjns replied:
For reference, see this commit: https://github.com/backstage-technical-services/website/commit/682ff457afc2a52d617b32a5e2d72cc62c83ba05

bnjns commented on resources/views/backups/modal.blade.php:
```diff
@@ -21,7 +21,7 @@
         <p class="help-block">
             A <strong>Database and Resources</strong> backup includes any files that aren't version controlled.
-            At the moment this includes: member profile pictures, resources and election manifestos.
+            At the moment this includes: member profile pictures, resources, breakage report images and election manifestos.
```
Is this correct? It probably should, but you'll need to configure it: https://github.com/backstage-technical-services/website/blob/main/config/backup.php#L29

bnjns commented on resources/views/equipment/repairs/view.blade.php:
```diff
@@ -56,6 +56,26 @@
             </div>
         </div>
 
+        <div class="form-group">
+            {!! Form::label('images', 'Images:', ['class' => 'control-label col-md-4']) !!}
+            <div class="col-md-8">
+                @if($breakage->images->count())
+                    <div class="row">
+                        @foreach($breakage->images as $image)
+                            <div class="col-xs-6 col-md-4">
+                                <a href="{{ $image->getImageRoute() }}" class="thumbnail" target="_blank">
+                                    <img src="{{ $image->getImageRoute() }}" alt="Image">
+                                </a>
+                            </div>
+                        @endforeach
+                    </div>
+                @else
+                    <p class="form-control-static"><em>- no images -</em></p>
+                @endif
+            </div>
+
+        </div>
```
We should probably wrap this entire form group in the if statement, so that we only show the entire thing if there are images. Also, I'd prefer a more explicit check so we don't have type coercion:

```blade
@if($breakage->images->count() > 0)
    ...
@endif
```

You could also move that to a helper function on the breakage if you wanted (eg, `hasImages`) but that's probably overkill.

bnjns replied:
Hmm, actually I'm less sure about this - definitely want the stricter check though!

bnjns commented on app/Http/Controllers/Equipment/RepairsController.php:
```diff
@@ -153,4 +168,26 @@ private function updateBreakageStatus(Breakage $breakage, Request $request)
+        $breakage = Breakage::findOrFail($id);
+        $this->authorize('view', $breakage);
+
+        $image = $breakage->images()->where('position_id', $imageId)->firstOrFail();
```
I'm sorta in 2 minds, but I feel like this can just be the image's PK rather than the position, but I'm open to debate :slightly_smiling_face:

bnjns commented on .docker/prod.Dockerfile:
```diff
@@ -36,6 +36,7 @@ COPY --chown=www-data:www-data --from=builder /var/www/vendor/ ./vendor/
 
 #  Declare the volumes
 VOLUME /var/www/public/images/profiles
+VOLUME /var/www/resources/breakages
```
Do we need to add a volume mount to the `docker-compose.yaml` file (and update the `local.Dockerfile` file) so that this directory is also kept for local dev?

Ben754444 replied:
It should persist because the whole of `/var/www` is a volume in `local.Dockerfile`

bnjns reviewed (changes requested):
Thanks so much for the PR - really appreciate you having a go at implementing this! Apologies it took so long to review :grimacing: 

In case you're not aware, comments that start with `nit:` are "nitpicking" - they're small comments/recommendations that don't necessarily need to be actioned.

Ben754444 commented:
Thanks for the feedback! I've removed `position_id` entirely, so everything should reference the primary key now. The filename has become the primary key, but we still need to know the extension so I've added a field for that. 

I think it's good to note that a breakage has no images rather than hiding the field but open to changing that! Might be worth changing the message from `no images` to `none`, though.

bnjns commented:
> I think it's good to note that a breakage has no images rather than hiding the field but open to changing that! Might be worth changing the message from `no images` to `none`, though.

Yeah, thinking about it I think you're right :slightly_smiling_face: I don't have any strong opinions either way on the wording, so I'll leave that to you

bnjns commented on app/Http/Controllers/Equipment/RepairsController.php:
```diff
@@ -153,4 +166,27 @@ private function updateBreakageStatus(Breakage $breakage, Request $request)
+        $breakage = Breakage::findOrFail($id);
+        $this->authorize('view', $breakage);
+
+        $image = BreakageImage::findOrFail($imageId);
```
This is a very minor comment, but I'd still recommend using something like `$breakage->images()->where('id', $imageId)->firstOrFail()` - it seems overkill but it makes sure that someone doesn't try to fetch an image for a different breakage.

It's not particularly problematic in this instance, but if we restricted who could access specific breakages this would allow someone to bypass that authorisation check for the images :slightly_smiling_face:

Election manifestos is an example where we do something similar: https://github.com/backstage-technical-services/website/blob/main/app/Http/Controllers/Elections/NominationController.php#L82

bnjns reviewed (approved):
Awesome stuff :tada:
